core/base_service.go

```diff
@@ -42,32 +41,38 @@ const (
 	URL                          = "url"
 	USERNAME                     = "username"
 	PASSWORD                     = "password"
-	IAM_API_KEY                  = "iam_apikey"
+	IAM_APIKEY                   = "iam_apikey"
 	IAM_URL                      = "iam_url"
 	SDK_NAME                     = "ibm-go-sdk-core"
 	UNKNOWN_ERROR                = "Unknown Error"
+	ICP4D                        = "icp4d"
+	IAM                          = "iam"
 )
 
 // ServiceOptions Service options
 type ServiceOptions struct {
-	Version         string
-	URL             string
-	Username        string
-	Password        string
-	IAMApiKey       string
-	IAMAccessToken  string
-	IAMURL          string
-	IAMClientId     string
-	IAMClientSecret string
+	Version            string
+	URL                string
+	Username           string
+	Password           string
+	IAMApiKey          string
+	IAMAccessToken     string
+	IAMURL             string
+	IAMClientId        string
+	IAMClientSecret    string
+	ICP4DAccessToken   string
+	ICP4DURL           string
+	AuthenticationType string
 }
 
 // BaseService Base Service
 type BaseService struct {
-	Options        *ServiceOptions
-	DefaultHeaders http.Header
-	TokenManager   *TokenManager
-	Client         *http.Client
-	UserAgent      string
+	Options           *ServiceOptions
+	DefaultHeaders    http.Header
+	IAMTokenManager   *IAMTokenManager
+	ICP4DTokenManager *ICP4DTokenManager
```

Comment:
For the next major release of the SDKs (September?) we'll want to move forward with a more abstract notion of authentication and (as much as is possible at least) separate out the authentication component from the core, or at least organize the code a little better to reduce coupling as much as we can.   With that in mind, it would be good if we could start moving in that direction and follow more of the model used in the java core.   You could create an interface that represents an "Authenticator" and there could be implementations such as "BasicAuthenticator", IAMAuthenticator (what is now IAMTokenManager) and ICP4DAuthenticator (the new token manager for icp4d).   I'll leave that up to you as to whether we want to start to move in that direction now or later :)

Specifically in this spot, we'd have a single pointer to an "Authenticator" which would support either basic auth, iam or icp4d, rather than having pointers to both an IAMTokenManager and an ICP4DTokenManager.

Reply:
After talking to @padamstx, we decided to go with the current format in the interest of time.